pkg/services/accesscontrol/seeder/seeder.go

```diff
@@ -0,0 +1,219 @@
+			},
+			{
+				Permission: "users.tokens:list",
+				Scope:      "users:self",
+			},
+			{
+				Permission: "users.teams:read",
+				Scope:      "users:self",
+			},
+		},
+	},
+}
+
+// FIXME: Make sure builtin grants can be removed without being recreated
+var builtInRoleGrants = map[string][]string{
+	"grafana:builtin:users:read:self": {
+		"Viewer",
+	},
+}
+
+func NewSeeder(s accesscontrol.AccessControl, log log.Logger) *seeder {
+	return &seeder{Store: s, log: log}
+}
+
+func (s *seeder) Seed(ctx context.Context, orgID int64) error {
+	err := s.seed(ctx, orgID, builtInRoles, builtInRoleGrants)
+	return err
+}
+
+func (s *seeder) seed(ctx context.Context, orgID int64, roles []accesscontrol.RoleDTO, roleGrants map[string][]string) error {
+	// FIXME: As this will run on startup, we want to optimize running this
+	existingRoles, err := s.Store.GetRoles(ctx, orgID)
+	if err != nil {
+		return err
+	}
+	roleSet := map[string]*accesscontrol.Role{}
+	for _, role := range existingRoles {
+		if role == nil {
+			continue
+		}
+		roleSet[role.Name] = role
+	}
+
+	for _, role := range roles {
+		role.OrgID = orgID
+
+		current, exists := roleSet[role.Name]
+		if exists {
+			if role.Version <= current.Version {
+				continue
+			}
+		}
+
+		roleID, err := s.createOrUpdateRole(ctx, role, current)
+		if err != nil {
+			s.log.Error("failed to create/update role", "name", role.Name, "err", err)
+			continue
+		}
+
+		if builtinRoles, exists := roleGrants[role.Name]; exists {
+			for _, builtinRole := range builtinRoles {
+				err := s.Store.AddBuiltinRole(ctx, orgID, roleID, builtinRole)
+				if err != nil && !errors.Is(err, accesscontrol.ErrUserRoleAlreadyAdded) {
+					s.log.Error("failed to assign role to role",
+						"name", role.Name,
+						"role", builtinRole,
+						"err", err,
+					)
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *seeder) createOrUpdateRole(ctx context.Context, role accesscontrol.RoleDTO, old *accesscontrol.Role) (int64, error) {
+	if role.Version == 0 {
+		return 0, fmt.Errorf("error when seeding '%s': all seeder roles must have a version", role.Name)
+	}
+
+	if old == nil {
+		p, err := s.Store.CreateRoleWithPermissions(ctx, accesscontrol.CreateRoleWithPermissionsCommand{
+			OrgID:       role.OrgID,
+			Version:     role.Version,
+			Name:        role.Name,
+			Description: role.Description,
+			Permissions: role.Permissions,
+		})
+		if err != nil {
+			return 0, err
+		}
+		return p.ID, nil
+	}
+
+	_, err := s.Store.UpdateRole(ctx, accesscontrol.UpdateRoleCommand{
+		UID:         old.UID,
+		Name:        role.Name,
+		Description: role.Description,
+		Version:     role.Version,
+	})
+	if err != nil {
+		if errors.Is(err, accesscontrol.ErrVersionLE) {
+			return old.ID, nil
+		}
+		return 0, err
+	}
+
+	existingPermissions, err := s.Store.GetRolePermissions(ctx, old.ID)
+	if err != nil {
+		s.log.Info("failed to get current permissions for role", "name", role.Name, "err", err)
```

Comment:
Is it OK to continue here?

Reply:
@sakjur ?

Reply:
🤔 I think both of these should be returned. I'll change that.